### PR TITLE
Put a safety limit on parallelization of tests

### PR DESCRIPTION
**Pull request by apolcyn**
... this is really just one idea for a solution:

https://github.com/grpc/grpc/pull/12774 slightly improves the performance of the port server, but it looks like that wasn't enough, as can be seen by the one port server request failure in [one of those windows/kokoro tests](https://sponge.corp.google.com/target?id=e3d4d3e1-42f5-4ccd-93d7-243f0695981a&target=github/grpc/c_windows_dbg_native&searchFor=&show=FAILED&sortBy=STATUS).

It appears that the parallelization of tests in terms of CPU cost is causing too many tests to run in parallel, to the point that the port server is getting overloaded (don't have data for this right now besides error codes but I suspect that the focusing of port server failures on windows is related to differing behavior between windows and linux when listen socket queues overrun - as I understand, for [windows](https://tangentsoft.net/wskfaq/advanced.html) and [linux](http://veithen.github.io/2014/01/01/how-tcp-backlog-works-in-linux.html)). (also in some experiments on kokoro windows, sometimes the run_tests.py script will fail with "too many open files" too). It's looking like this large parallelization is the main cause of https://github.com/grpc/grpc/issues/12595.

With this change combined with https://github.com/grpc/grpc/pull/12774, I'm seeing windows core tests consistently pass (no port server related failures) on experimental kokoro windows "pull-request" jobs.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: tools/run_tests/run_tests.py [hunk 1]
```diff
@@ -57,6 +57,9 @@
 _ROOT = os.path.abspath(os.path.join(os.path.dirname(sys.argv[0]), '../..'))
 os.chdir(_ROOT)
 
+# Avoid running too many tests in parrallel (e.g., to not overload port server).
+# TODO: possibly tune this.
+_MAX_PARALLEL_TESTS_CPU_AGNOSTIC = 32
```
Comment (jtattermusch) [inline tools/run_tests/run_tests.py]: once we start using 32core and 64core workers, the default of 32 seems insufficient. 
Would things still work fine with _MAX_PARALLEL_TESTS_CPU_AGNOSTIC = 64?

**Comment by apolcyn** [comment]
@jtattermusch just updated to make the limits platform specific. I'm not aware of port-server/test-env-scaling issues outside of windows so I removed those limits for now. Experimenting to see if a 64 limit on windows is ok.

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by apolcyn** [comment]
at least on a kokoro windows PR experiment job, a limit of 64 for windows seems ok so far

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by apolcyn** [comment]
Summarizing some of the test durations from the last run for comparison:

Might be worth doing another run too...

* `linux c++ dbg kokoro`:
1.5 sec waiting in the queue;
1 hr 7 min building on an executor;
1 hr 7 min total from scheduled to completion.
  * [most recent run](https://kokoro.corp.google.com/job/grpc/job/ubuntu/job/pull_request/job/grpc_basictests_cpp_dbg/397/), ran in 1hr 5 min

* mac tests didn't run (failed with `coreutils already installed` error, might re-run these)

* multi-lang linux tests (seems a little long compared to usual ~50 minutes which I'm not sure why)
1.3 sec waiting in the queue;
1 hr 46 min building on an executor;
1 hr 46 min total from scheduled to completion.
  * [most recent run, not from this PR](https://kokoro.corp.google.com/job/grpc/job/ubuntu/job/pull_request/job/grpc_basictests_multilang/2685/)

* windows dbg kokoro tests (with no port server failures :) )
0.24 sec waiting in the queue;
59 min building on an executor;
59 min total from scheduled to completion.
  * [most recent run, not from this PR](https://kokoro.corp.google.com/job/grpc/job/windows/job/pull_request/job/grpc_basictests_dbg/262/) ran in 54 minutes.

* windows opt kokoro tests
1.7 sec waiting in the queue;
31 min building on an executor;
31 min total from scheduled to completion.
  * [most recent run](https://sponge.corp.google.com/target?id=7564965d-fb07-4205-908a-df4148adea67&target=github/grpc/c_windows_opt_native&searchFor=&show=FAILED&sortBy=STATUS), ran in 42 minutes.

* basic test c dbg kokoro
1.4 sec waiting in the queue;
11 min building on an executor;
11 min total from scheduled to completion.
  * [most recent run](https://kokoro.corp.google.com/job/grpc/job/ubuntu/job/pull_request/job/grpc_basictests_c_dbg/365/) ran in 11 minutesc

**Comment by apolcyn** [comment]
[timeout seen](https://sponge.corp.google.com/target?id=fd0fa528-8f5d-422f-8286-bae6644a3972&target=grpc/ubuntu/pull_request/grpc_basictests_multilang&searchFor=&show=FAILED&sortBy=STATUS) in `tools/run_tests/helper_scripts/run_ruby_end2end_tests.sh` (this is usually a < 1 minute test, that surpassed it's 10 minute timeout. But it doesn't appear related... re-running basic tests multi-lang

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by apolcyn** [comment]
test failures: (noting some times here too)
* c++ linux basic tests dbg kokoro: https://github.com/grpc/grpc/issues/12510
  - 1 hr. 12 min, looks normal
* c++ linux basic tests opt kokoro: https://github.com/grpc/grpc/issues/12510, https://github.com/grpc/grpc/issues/11737
  - 1 hr., 8 min, looks normal
* windows dbg basic tests kokoro: https://github.com/grpc/grpc/issues/11109, https://github.com/grpc/grpc/issues/12812
  - 58 min., looks normal
* https://github.com/grpc/grpc/issues/9037
  - windows opt basic tests kokoro: https://github.com/grpc/grpc/issues/9037
  - 45 min., looks normal
* interop cloud to prod: https://github.com/grpc/grpc/issues/10763
* msan C tests: https://github.com/grpc/grpc/issues/12394
  - 41 min., looks typical

**Comment by jtattermusch** [comment]
Thanks!

**Pull request merged**